File: src/test/java/org/simbrain/network/synapse_update_rules/spikeresponders/SpikeResponderTest.java
package org.simbrain.network.synapse_update_rules.spikeresponders;

import org.junit.Test;
import org.simbrain.network.core.Network;
import org.simbrain.network.core.Neuron;
import org.simbrain.network.core.Synapse;
import org.simbrain.network.neuron_update_rules.IzhikevichRule;
import org.simbrain.network.neuron_update_rules.LinearRule;

import static org.junit.Assert.*;

public class SpikeResponderTest {


    @Test
    public void basicTest() {

        Network net = new Network();
        IzhikevichRule ir = new IzhikevichRule();
        ir.setiBg(15);
        Neuron spiking = new Neuron(net, ir);
        LinearRule lr = new LinearRule();
        lr.setClipped(false);
        Neuron output = new Neuron(net, lr );
        Synapse s = new Synapse(spiking, output);
        net.addNeuron(spiking);
        net.addNeuron(output);

        for (int i = 0; i < 100 ; i++) {
            net.update();
        }

        // TODO: Figure out expected number of spikes


    }




}

File: src/test/kotlin/org/simbrain/network/core/SpikeResponderTest.kt
package org.simbrain.network.core

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.simbrain.network.neuron_update_rules.SpikingThresholdRule
import org.simbrain.network.synapse_update_rules.spikeresponders.JumpAndDecay
import org.simbrain.network.synapse_update_rules.spikeresponders.Step

class SpikeResponderTest {

    val net = Network()
    val rule = SpikingThresholdRule()
    val n1 = Neuron(net, rule)
    val n2 = Neuron(net)
    val s = Synapse(n1, n2)

    @BeforeEach
    fun setup() {
        net.timeStep = 1.0 // For simpler computations
        net.addNetworkModels(listOf(n1, n2, s))
    }

    @Test
    fun `test step responder`() {
        rule.threshold = .7
        n2.upperBound = 10.0

        val step = Step()
        step.responseHeight = 2.0
        step.responseDuration = 3.0
        s.spikeResponder = step
        net.timeStep = 1.0 // For simpler computations
        net.addNetworkModels(listOf(n1, n2, s))
        println(s.spikeResponder)

        // Sub-threshold.
        n1.addInputValue(.6)
        net.update()
        assertEquals(0.0, n2.activation)

        // Above-threshold. Should "fire" at responseHeight (2.0) for responseDuration (3)
        n1.addInputValue(.8)
        net.update()
        net.update()  // TODO: Why does it take an extra step?
        assertEquals(2.0, n2.activation)
        net.update()
        assertEquals(2.0, n2.activation)
        net.update()
        assertEquals(2.0, n2.activation)
        net.update()
        assertEquals(0.0, n2.activation)

        // repeat(10) {
        //     if (it == 3) n1.addInputValue(1.0)
        //     net.update()
        //     println("$it:${n1.activation} -> ${n2.activation}")
        // }
    }

    @Test
    fun `test jump and decay`() {
        rule.threshold = .7
        n2.upperBound = 10.0

        val sr = JumpAndDecay()
        s.spikeResponder = sr

        repeat(10) {
            if (it == 3) n1.addInputValue(1.0)
            net.update()
            println("$it:${n1.activation} -> ${n2.activation}")
        }
    }

}